### PR TITLE
Docker Swarm HA Gateway [NB2KG dependency on passing token]

### PR DESCRIPTION
**Pull request by mattjtodd**
This is a follow-on Issue #632 It makes sense to me to separate out the two at this time.

I've looked at the FIXME issues with Docker Swarm HA mode for the gateway and how it could be resolved.

Session stickiness is handled by a dynamic cookie which when set by clients after the first request results in a consistent replicated instance being used by proxy.  If this type of affinity is not possible, then I can explore other ways.  As a starter for ten, this is an example of how `dnsrr` and docker swarm mesh routing is designed to work together with HAProxy to get HA and session affinity.

The FixMe points of interest in the swarm script is:

```
# 2. Using endpoint-mode `dnsrr` (which appears to be required in-order for kernel container to send the connection info response back) 
# also required mode=host on any published ports. :-(
# 3. We only use one replica since session affinity is another point of investigation in Swarm
```

2. DNSRR is the correct load balancing protocol for adding a custom load balancer in-front of a docker DNS service registry in this case as it provides an A record of containers ips.  This can be managed by HAProxy resolvers to dynamically scale.  Host mode is not required as a reverse proxy such as HAProxy can publish an ingress port and provide forwarding to the private services yielding the desired scheduler elasticity.

3. Session affinity in this example is handled using a dynamic session cookie which is part of [HAProxy's DNS service discovery mechanism](https://www.haproxy.com/blog/dns-service-discovery-haproxy/).   If this is not a suitable affinity mechanism, then alternatives could be investigated.

HAProxy has been configured to expose the stats UI, which can be accessed on port `:9088`, so locally running cluster:

http://localhost:9088/

Not that this configuration doesn't consider HA / replicated HAProxy configuration.

### CODE REVIEW

File: etc/docker/docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+          timeout client 5000
+          default_backend enterprise-gateway
+
+        backend enterprise-gateway
+          mode http
+          timeout connect 100
+          timeout server 5000
+          dynamic-cookie-key ENTERPRISE_KEY
+          cookie SRVID insert dynamic
+          server-template enterprise-gateway 5 enterprise-gateway:8888 check resolvers docker
```
[the remark this answers is not in the document]
Comment (mattjtodd) [reply etc/docker/docker-compose.yml]: Will do.

**Comment by kevin-bates** [comment]
Hi Matt - This PR no longer resolves because it had referenced commits from #633 and is now merged into master.  Please resolve the conflicts and we'll get this merged afterward.  Thanks.

**Comment by mattjtodd** [comment]
I have another change to address an issue with networking before this is ready to go in, I'll get it in today.

**Comment by kevin-bates** [comment]
Thanks Matt.  I see a couple commits today.  Are you expecting further updates?

**Comment by kevin-bates** [comment]
Thanks Matt.  I see a couple commits today.  Are you expecting further updates?

**Comment by mattjtodd** [comment]
Possibly, I'm trying to debug a case where the `nb2kg` isn't successfully being sticky.  Without replication of the `enterprise-gateway` everything works fine.  It's likely that the websocket connection isn't adhering to the affinity.

File: etc/docker/docker-compose.yml [hunk 2]
```diff
@@ -43,5 +38,71 @@ services:
...
+          log global
+          option httplog
+          mode http
+          option  http-server-close
+          option  dontlognull
+          option  redispatch
+          option  contstats
+          retries 3
+          backlog 10000
+          timeout client          25s
```
Comment (esevan) [inline etc/docker/docker-compose.yml]: Question. these `timeout` value don't close websocket connection?

**Comment by mattjtodd** [comment]
I've spent some more time looking into this and the functionality from the proxy is correct, which auto-generates a cookie which can be used to create sticky connections to the backend-instances.  I've sniffed the traffic over the wire using `tshark` and can confirm that although the cookie is received by the browser subsequent requests via the `nb2kg` service shows that the cookie is not being passed back to the gateway.

It may be worth considering an architectural change to not persist the state in the gateway, but to delegate this a KV service such as [Consul](https://www.consul.io/), [Zookeeper](https://zookeeper.apache.org/) or similar.

I've now added all my changes, thanks for reviewing.

**Comment by kevin-bates** [comment]
Thanks for digging so deeply!  However, based on your response, it's not entirely clear to me how we should proceed.  Unfortunately, I'm not very knowledgable in this web-interface space - so I'm hoping others can chime in here.  

Does merging these changes only yield something that is no longer usable?  
Are there things we need to look into NB2KG about?

When you say...
>It may be worth considering an architectural change to not persist the state in the gateway, but to delegate this a KV service such as Consul, Zookeeper or similar.

what _state_ are you referring to that is persisted?  If it's the kernel session stuff, then, yes, I absolutely agree we should be persisting that state into a NoSQL DB - which I suspect you're getting at with a key-value service - this is what #594 is driving at.  Are you referring to something other than the kernel session state?

Please advise - thank you.

**Comment by lresende** [comment]
> I've spent some more time looking into this and the functionality from the proxy is correct, which auto-generates a cookie which can be used to create sticky connections to the backend-instances. I've sniffed the traffic over the wire using `tshark` and can confirm that although the cookie is received by the browser subsequent requests via the `nb2kg` service shows that the cookie is not being passed back to the gateway.

I believe NB2KG needs to be properly configured to submit/propagate specific tokens when talking to JEG. Maybe @kevin-bates can give more details here.

**Comment by lresende** [comment]
> It may be worth considering an architectural change to not persist the state in the gateway, but to delegate this a KV service such as [Consul](https://www.consul.io/), [Zookeeper](https://zookeeper.apache.org/) or similar.

We should consider these type of abstractions for sure, but, as we have support for both containerized and non-containerized environments, we should continue to support both which might cause us to have two implementations and different defaults for different environments.

**Comment by mattjtodd** [comment]
@kevin-bates yes it's the JEG gateway state I'm referring to.  #594 is looking at solution, but it may be that a simple KV store could be simpler and provide a more suitable solution as it will handle locking and race conditions out of the box.  I've not had a chance to look into the session management code too closely yet and I'm not a python code so can't say for sure if it would be the best fit.  I'll try and get a chance to look in the nest week or so.

@lresende indeed, the cookie would need to be passed back to the JEG proxy for it to be able to use a sticky session and access the session stored in a given instance.

The choice of and use of an external persistent storage for the session state would not be affected by the use of containers, each of those technologies do not mandate their being run in one, so would remain portable between environments.  I like to think about containers being a thin veneer of portability covering a single linux process and it's required files which really shouldn't be driving architectural decisions from a logical level and represents a significant benefit to portability and deployability options.

I'd be really interested to help looking into how the session could be moved to some storage which would increase complexity in terms of design, but would greatly increase the resilience of the JEG sessions with the added benefit of handing off the complexity of HA state management to another established service.

If I can get to the point where I can understand the state management requirements and python I should be able to get a design and POC together.

**Comment by kevin-bates** [comment]
@mattjtodd - thanks for the response.  Could you be more specific as to the _state_ you're referring to regarding persistence?  To my knowledge, the only state we need to worry about is the state used to reestablish connections to running kernels - which I refer to as _kernel session_ state.

When enabled (off by default), this state is currently maintained by the [`KernelSessionManager`](https://github.com/jupyter/enterprise_gateway/blob/master/enterprise_gateway/services/sessions/kernelsessionmanager.py#L20).  When a kernel is started, just prior to the response back to the client, the state is captured and saved.  When a kernel is shutdown, the state is destroyed.  Whenever EG starts, it checks the store for existing sessions and attempts to "hydrate" the KernelManager class instance with the persisted information.  This comment contains some further information for Active/Active: https://github.com/jupyter/enterprise_gateway/issues/562#issuecomment-458203989

To your point regarding KV stores, we need to implement better abstraction in the KernelSessionManager's load and save methods such that they utilize the abstraction rather a file directly.

**Comment by kevin-bates** [comment]
@mattjtodd - where are we with this PR?  I believe you were going to look into something related to nb2kg...
>I've sniffed the traffic over the wire using tshark and can confirm that although the cookie is received by the browser subsequent requests via the nb2kg service shows that the cookie is not being passed back to the gateway.

From above:
>Does merging these changes only yield something that is no longer usable?

Just want to be sure if we took this change what doesn't work?  Do things fail only when attempting to use multiple EGs simultaneously - or, with this change, EG doesn't work at all?

**Comment by mattjtodd** [comment]
@kevin-bates sorry I've been unable to look at this work for a while, apologies.

You're right, this problem only arises when there is no cookie and the EG replicas > 1.  The state stored by the EG for connecting to the kernels is what I believe is causing the issue, move that somewhere else and make the EG stateless or use the cookie from the nb2kg and this will allow for HA.

For a single EG the solution works as expected.

**Comment by kevin-bates** [comment]
@mattjtodd - Thank you for the response.  

I see.  So we get session affinity in Swarm via a cookie from the client (nb2kg or other)?   I think there will be issues with Active/Active operations in its absence - which I allude to [here](https://github.com/jupyter/enterprise_gateway/issues/594#issuecomment-485968074) - so true statelessness is probably a ways away.

**Pull request merged**